File: frontend/src/types/financing.ts
// frontend/src/types/financing.ts
/** Financing module types — derived from backend schemas (backend/schemas/financing.py). */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export type FinancingInstrumentType =
  | 'conventional_mortgage'
  | 'sub_to_mortgage'
  | 'seller_finance'
  | 'wrap_mortgage'
  | 'lease_option'
  | 'hard_money'
  | 'private_money'
  | 'heloc'
  | 'land_contract'

export type ObligationType =
  | 'monthly_payment'
  | 'balloon_payment'
  | 'insurance_renewal'
  | 'option_expiration'
  | 'rate_adjustment'
  | 'manual'

export type ObligationStatus = 'active' | 'completed' | 'snoozed'
export type ObligationSeverity = 'normal' | 'high' | 'critical'
export type PaymentDirection = 'outgoing' | 'incoming'
export type RateType = 'fixed' | 'adjustable' | 'interest_only'

export type PaymentType =
  | 'regular'
  | 'extra_principal'
  | 'balloon'
  | 'late_fee'
  | 'insurance'
  | 'tax'

export type PaymentMethod = 'bank_transfer' | 'check' | 'cash' | 'auto_pay'

// ---------------------------------------------------------------------------
// Core entities
// ---------------------------------------------------------------------------

export interface FinancingInstrument {
  id: string
  property_id: string
  deal_id: string | null
  created_by: string
  team_id: string | null
  name: string
  instrument_type: string
  position: number
  status: string

  // Core terms
  original_balance: number | null
  current_balance: number | null
  interest_rate: number | null
  rate_type: string | null
  term_months: number | null
  amortization_months: number | null
  monthly_payment: number | null

  // Dates
  origination_date: string | null
  maturity_date: string | null
  first_payment_date: string | null

  // Balloon
  has_balloon: boolean
  balloon_date: string | null
  balloon_amount: number | null

  // Sub-to
  is_sub_to: boolean
  original_borrower: string | null
  servicer: string | null
  loan_number_last4: string | null
  due_on_sale_risk: string | null

  // Wrap
  is_wrap: boolean
  underlying_instrument_id: string | null
  wrap_rate: number | null
  wrap_payment: number | null

  // Lease option
  option_consideration: number | null
  option_expiration: string | null
  monthly_credit: number | null
  strike_price: number | null

  // Seller finance
  down_payment: number | null
  late_fee_pct: number | null
  late_fee_grace_days: number | null
  prepayment_penalty: boolean

  // Insurance/escrow
  requires_insurance: boolean
  insurance_verified_date: string | null
  escrow_amount: number | null

  // Extended
  terms_extended: Record<string, unknown> | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface InstrumentListItem extends FinancingInstrument {
  months_remaining: number | null
  next_payment_due: string | null
  payoff_amount_estimate: number | null
  wrap_spread: WrapSpreadSummary | null
}

export interface InstrumentDetail extends FinancingInstrument {
  obligations: Obligation[]
  recent_payments: Payment[]
  wrap_spread: WrapSpreadSummary | null
  amortization_schedule: AmortizationEntry[]
}

export interface PaginatedInstruments {
  items: InstrumentListItem[]
  total: number
  page: number
  per_page: number
  pages: number
}

// ---------------------------------------------------------------------------
// Obligation
// ---------------------------------------------------------------------------

export interface Obligation {
  id: string
  instrument_id: string
  property_id: string
  created_by: string
  team_id: string | null
  obligation_type: string
  title: string
  description: string | null
  amount: number | null
  amount_type: string | null
  due_date: string | null
  recurrence: string | null
  recurrence_day: number | null
  next_due: string | null
  end_date: string | null
  status: string
  alert_days_before: number[] | null
  severity: string
  created_at: string
  updated_at: string
  // Computed fields (from ObligationWithComputed)
  days_until_due: number | null
  is_overdue: boolean
  instrument_name: string | null
  property_address: string | null
}

export interface ObligationGrouped {
  critical: Obligation[]
  high: Obligation[]
  normal: Obligation[]
}

export interface PaginatedObligations {
  items: Obligation[]
  total: number
  page: number
  per_page: number
  pages: number
}

// ---------------------------------------------------------------------------
// Payment
// ---------------------------------------------------------------------------

export interface Payment {
  id: string
  instrument_id: string
  obligation_id: string | null
  property_id: string
  created_by: string
  team_id: string | null
  payment_type: string
  amount: number
  principal_portion: number | null
  interest_portion: number | null
  escrow_portion: number | null
  payment_date: string
  due_date: string | null
  is_late: boolean
  late_fee_amount: number | null
  payment_method: string | null
  confirmation_number: string | null
  direction: string
  notes: string | null
  created_at: string
  updated_at: string
}

export interface PaginatedPayments {
  items: Payment[]
  total: number
  page: number
  per_page: number
  pages: number
}

// ---------------------------------------------------------------------------
// Wrap Spread
// ---------------------------------------------------------------------------

export interface WrapSpreadSummary {
  monthly_incoming: number
  monthly_outgoing: number
  monthly_spread: number
  annual_spread: number
  spread_margin_pct: number
  underlying_rate: number
  wrap_rate: number
  rate_spread: number
}

// ---------------------------------------------------------------------------
// Amortization
// ---------------------------------------------------------------------------

export interface AmortizationEntry {
  month: number
  payment_date: string | null
  payment: number
  principal: number
  interest: number
  balance: number
}

// ---------------------------------------------------------------------------
// Dashboard
// ---------------------------------------------------------------------------

export interface BalloonAlert {
  instrument_name: string
  property_address: string
  balloon_date: string | null
  balloon_amount: number | null
  days_until: number | null
}

export interface WrapSpreadItem {
  property_address: string
  monthly_spread: number
  annual_spread: number
}

export interface DueOnSaleRiskItem {
  property_address: string
  instrument_name: string
  risk_level: string
}

export interface FinancingDashboard {
  total_instruments: number
  total_balance: number
  total_monthly_obligations: number
  total_monthly_income: number
  net_monthly_cash_flow: number
  upcoming_balloons: BalloonAlert[]
  wrap_spreads: WrapSpreadItem[]
  due_on_sale_risks: DueOnSaleRiskItem[]
}

// ---------------------------------------------------------------------------
// Payment Summary
// ---------------------------------------------------------------------------

export interface PropertyPaymentSummary {
  property_id: string
  address: string
  total: number
}

export interface MonthlyHistory {
  month: string
  incoming: number
  outgoing: number
  net: number
}

export interface PaymentSummaryResponse {
  total_outgoing_monthly: number
  total_incoming_monthly: number
  net_monthly: number
  outgoing_by_property: PropertyPaymentSummary[]
  incoming_by_property: PropertyPaymentSummary[]
  payment_history: MonthlyHistory[]
}

// ---------------------------------------------------------------------------
// Request types
// ---------------------------------------------------------------------------

export interface CreateInstrumentRequest {
  property_id: string
  deal_id?: string
  name: string
  instrument_type: string
  position?: number
  status?: string
  original_balance?: number
  current_balance?: number
  interest_rate?: number
  rate_type?: string
  term_months?: number
  amortization_months?: number
  monthly_payment?: number
  origination_date?: string
  maturity_date?: string
  first_payment_date?: string
  has_balloon?: boolean
  balloon_date?: string
  balloon_amount?: number
  is_sub_to?: boolean
  original_borrower?: string
  servicer?: string
  loan_number_last4?: string
  due_on_sale_risk?: string
  is_wrap?: boolean
  underlying_instrument_id?: string
  wrap_rate?: number
  wrap_payment?: number
  option_consideration?: number
  option_expiration?: string
  monthly_credit?: number
  strike_price?: number
  down_payment?: number
  late_fee_pct?: number
  late_fee_grace_days?: number
  prepayment_penalty?: boolean
  requires_insurance?: boolean
  insurance_verified_date?: string
  escrow_amount?: number
  terms_extended?: Record<string, unknown>
  notes?: string
}

export interface UpdateInstrumentRequest {
  [key: string]: unknown
}

export interface CompleteObligationRequest {
  payment_amount?: number
  payment_date?: string
  payment_method?: string
}

export interface UpdateObligationRequest {
  title?: string
  description?: string
  amount?: number
  next_due?: string
  status?: string
  severity?: string
  alert_days_before?: number[]
}

export interface CreatePaymentRequest {
  instrument_id: string
  obligation_id?: string
  property_id: string
  payment_type: string
  amount: number
  principal_portion?: number
  interest_portion?: number
  escrow_portion?: number
  payment_date: string
  due_date?: string
  is_late?: boolean
  late_fee_amount?: number
  payment_method?: string
  confirmation_number?: string
  direction?: string
  notes?: string
}

// ---------------------------------------------------------------------------
// Filter types
// ---------------------------------------------------------------------------

export interface InstrumentFilters {
  property_id?: string
  status?: string
  instrument_type?: string
  page?: number
  per_page?: number
}

export interface ObligationFilters {
  instrument_id?: string
  property_id?: string
  status?: string
  severity?: string
  due_before?: string
  due_after?: string
  page?: number
  per_page?: number
}

export interface PaymentFilters {
  instrument_id?: string
  property_id?: string
  direction?: string
  date_from?: string
  date_to?: string
  page?: number
  per_page?: number
}

File: backend/schemas/financing.py
"""Pydantic schemas for financing instruments, obligations, and payments."""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# FinancingInstrument
# ---------------------------------------------------------------------------

class CreateFinancingInstrumentRequest(BaseModel):
    property_id: UUID
    deal_id: Optional[UUID] = None
    name: str = Field(..., min_length=1, max_length=200)
    instrument_type: str
    position: int = 1
    status: str = "active"

    # Core terms
    original_balance: Optional[Decimal] = None
    current_balance: Optional[Decimal] = None
    interest_rate: Optional[Decimal] = None
    rate_type: Optional[str] = None
    term_months: Optional[int] = None
    amortization_months: Optional[int] = None
    monthly_payment: Optional[Decimal] = None

    # Dates
    origination_date: Optional[date] = None
    maturity_date: Optional[date] = None
    first_payment_date: Optional[date] = None

    # Balloon
    has_balloon: bool = False
    balloon_date: Optional[date] = None
    balloon_amount: Optional[Decimal] = None

    # Sub-to
    is_sub_to: bool = False
    original_borrower: Optional[str] = None
    servicer: Optional[str] = None
    loan_number_last4: Optional[str] = Field(None, max_length=4)
    due_on_sale_risk: Optional[str] = None

    # Wrap
    is_wrap: bool = False
    underlying_instrument_id: Optional[UUID] = None
    wrap_rate: Optional[Decimal] = None
    wrap_payment: Optional[Decimal] = None

    # Lease option
    option_consideration: Optional[Decimal] = None
    option_expiration: Optional[date] = None
    monthly_credit: Optional[Decimal] = None
    strike_price: Optional[Decimal] = None

    # Seller finance
    down_payment: Optional[Decimal] = None
    late_fee_pct: Optional[Decimal] = None
    late_fee_grace_days: Optional[int] = None
    prepayment_penalty: bool = False

    # Insurance/escrow
    requires_insurance: bool = True
    insurance_verified_date: Optional[date] = None
    escrow_amount: Optional[Decimal] = None

    # Extended
    terms_extended: Optional[dict] = None
    notes: Optional[str] = None


class UpdateFinancingInstrumentRequest(BaseModel):
    name: Optional[str] = None
    instrument_type: Optional[str] = None
    position: Optional[int] = None
    status: Optional[str] = None
    original_balance: Optional[Decimal] = None
    current_balance: Optional[Decimal] = None
    interest_rate: Optional[Decimal] = None
    rate_type: Optional[str] = None
    term_months: Optional[int] = None
    amortization_months: Optional[int] = None
    monthly_payment: Optional[Decimal] = None
    origination_date: Optional[date] = None
    maturity_date: Optional[date] = None
    first_payment_date: Optional[date] = None
    has_balloon: Optional[bool] = None
    balloon_date: Optional[date] = None
    balloon_amount: Optional[Decimal] = None
    is_sub_to: Optional[bool] = None
    original_borrower: Optional[str] = None
    servicer: Optional[str] = None
    loan_number_last4: Optional[str] = Field(None, max_length=4)
    due_on_sale_risk: Optional[str] = None
    is_wrap: Optional[bool] = None
    underlying_instrument_id: Optional[UUID] = None
    wrap_rate: Optional[Decimal] = None
    wrap_payment: Optional[Decimal] = None
    option_consideration: Optional[Decimal] = None
    option_expiration: Optional[date] = None
    monthly_credit: Optional[Decimal] = None
    strike_price: Optional[Decimal] = None
    down_payment: Optional[Decimal] = None
    late_fee_pct: Optional[Decimal] = None
    late_fee_grace_days: Optional[int] = None
    prepayment_penalty: Optional[bool] = None
    requires_insurance: Optional[bool] = None
    insurance_verified_date: Optional[date] = None
    escrow_amount: Optional[Decimal] = None
    terms_extended: Optional[dict] = None
    notes: Optional[str] = None


class FinancingInstrumentResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    property_id: UUID
    deal_id: Optional[UUID] = None
    created_by: UUID
    team_id: Optional[UUID] = None
    name: str
    instrument_type: str
    position: int
    status: str
    original_balance: Optional[Decimal] = None
    current_balance: Optional[Decimal] = None
    interest_rate: Optional[Decimal] = None
    rate_type: Optional[str] = None
    term_months: Optional[int] = None
    amortization_months: Optional[int] = None
    monthly_payment: Optional[Decimal] = None
    origination_date: Optional[date] = None
    maturity_date: Optional[date] = None
    first_payment_date: Optional[date] = None
    has_balloon: bool
    balloon_date: Optional[date] = None
    balloon_amount: Optional[Decimal] = None
    is_sub_to: bool
    original_borrower: Optional[str] = None
    servicer: Optional[str] = None
    loan_number_last4: Optional[str] = None
    due_on_sale_risk: Optional[str] = None
    is_wrap: bool
    underlying_instrument_id: Optional[UUID] = None
    wrap_rate: Optional[Decimal] = None
    wrap_payment: Optional[Decimal] = None
    option_consideration: Optional[Decimal] = None
    option_expiration: Optional[date] = None
    monthly_credit: Optional[Decimal] = None
    strike_price: Optional[Decimal] = None
    down_payment: Optional[Decimal] = None
    late_fee_pct: Optional[Decimal] = None
    late_fee_grace_days: Optional[int] = None
    prepayment_penalty: bool
    requires_insurance: bool
    insurance_verified_date: Optional[date] = None
    escrow_amount: Optional[Decimal] = None
    terms_extended: Optional[dict] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class InstrumentSummary(BaseModel):
    """Compact instrument view for lists."""
    model_config = {"from_attributes": True}

    id: UUID
    name: str
    instrument_type: str
    current_balance: Optional[Decimal] = None
    interest_rate: Optional[Decimal] = None
    monthly_payment: Optional[Decimal] = None
    status: str


# ---------------------------------------------------------------------------
# Obligation
# ---------------------------------------------------------------------------

class CreateObligationRequest(BaseModel):
    instrument_id: UUID
    property_id: UUID
    obligation_type: str
    title: str = Field(..., min_length=1)
    description: Optional[str] = None
    amount: Optional[Decimal] = None
    amount_type: Optional[str] = None
    due_date: Optional[date] = None
    recurrence: Optional[str] = None
    recurrence_day: Optional[int] = Field(None, ge=1, le=28)
    next_due: Optional[date] = None
    end_date: Optional[date] = None
    status: str = "active"
    alert_days_before: Optional[list[int]] = None
    severity: str = "normal"


class ObligationResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    instrument_id: UUID
    property_id: UUID
    created_by: UUID
    team_id: Optional[UUID] = None
    obligation_type: str
    title: str
    description: Optional[str] = None
    amount: Optional[Decimal] = None
    amount_type: Optional[str] = None
    due_date: Optional[date] = None
    recurrence: Optional[str] = None
    recurrence_day: Optional[int] = None
    next_due: Optional[date] = None
    end_date: Optional[date] = None
    status: str
    alert_days_before: Optional[list[int]] = None
    severity: str
    created_at: datetime
    updated_at: datetime

    @property
    def days_until_due(self) -> Optional[int]:
        if self.next_due:
            return (self.next_due - date.today()).days
        return None


class ObligationAlert(BaseModel):
    """Obligation surfaced in Today briefing."""
    id: UUID
    title: str
    due_date: Optional[date] = None
    days_until_due: Optional[int] = None
    severity: str
    instrument_name: str
    property_address: str


# ---------------------------------------------------------------------------
# Payment
# ---------------------------------------------------------------------------

class CreatePaymentRequest(BaseModel):
    instrument_id: UUID
    obligation_id: Optional[UUID] = None
    property_id: UUID
    payment_type: str
    amount: Decimal = Field(..., gt=0)
    principal_portion: Optional[Decimal] = Field(None, ge=0)
    interest_portion: Optional[Decimal] = Field(None, ge=0)
    escrow_portion: Optional[Decimal] = None
    payment_date: date
    due_date: Optional[date] = None
    is_late: bool = False
    late_fee_amount: Optional[Decimal] = None
    payment_method: Optional[str] = None
    confirmation_number: Optional[str] = None
    direction: str = "outgoing"
    notes: Optional[str] = None


class PaymentResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    instrument_id: UUID
    obligation_id: Optional[UUID] = None
    property_id: UUID
    created_by: UUID
    team_id: Optional[UUID] = None
    payment_type: str
    amount: Decimal
    principal_portion: Optional[Decimal] = None
    interest_portion: Optional[Decimal] = None
    escrow_portion: Optional[Decimal] = None
    payment_date: date
    due_date: Optional[date] = None
    is_late: bool
    late_fee_amount: Optional[Decimal] = None
    payment_method: Optional[str] = None
    confirmation_number: Optional[str] = None
    direction: str
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Wrap Spread
# ---------------------------------------------------------------------------

class WrapSpreadSummary(BaseModel):
    """Spread between wrap collections and underlying payments."""
    monthly_incoming: float
    monthly_outgoing: float
    monthly_spread: float
    annual_spread: float
    spread_margin_pct: float
    underlying_rate: float
    wrap_rate: float
    rate_spread: float


# ---------------------------------------------------------------------------
# Amortization
# ---------------------------------------------------------------------------

class AmortizationEntry(BaseModel):
    month: int
    payment_date: Optional[date] = Field(None, alias="date")
    payment: float
    principal: float
    interest: float
    balance: float

    model_config = {"populate_by_name": True}


# ---------------------------------------------------------------------------
# Instrument — list & detail views
# ---------------------------------------------------------------------------

class InstrumentListItem(FinancingInstrumentResponse):
    months_remaining: Optional[int] = None
    next_payment_due: Optional[date] = None
    payoff_amount_estimate: Optional[float] = None
    wrap_spread: Optional[WrapSpreadSummary] = None


class InstrumentDetailResponse(FinancingInstrumentResponse):
    obligations: list[ObligationResponse] = []
    recent_payments: list[PaymentResponse] = []
    wrap_spread: Optional[WrapSpreadSummary] = None
    amortization_schedule: list[AmortizationEntry] = []


class PaginatedInstruments(BaseModel):
    items: list[InstrumentListItem]
    total: int
    page: int
    per_page: int
    pages: int


# ---------------------------------------------------------------------------
# Obligation — computed, grouped, update, complete
# ---------------------------------------------------------------------------

class ObligationWithComputed(BaseModel):
    """Obligation with computed display fields (avoids @property shadow from parent)."""
    model_config = {"from_attributes": True}

    id: UUID
    instrument_id: UUID
    property_id: UUID
    created_by: UUID
    team_id: Optional[UUID] = None
    obligation_type: str
    title: str
    description: Optional[str] = None
    amount: Optional[Decimal] = None
    amount_type: Optional[str] = None
    due_date: Optional[date] = None
    recurrence: Optional[str] = None
    recurrence_day: Optional[int] = None
    next_due: Optional[date] = None
    end_date: Optional[date] = None
    status: str
    alert_days_before: Optional[list[int]] = None
    severity: str
    created_at: datetime
    updated_at: datetime
    # Computed
    days_until_due: Optional[int] = None
    is_overdue: bool = False
    instrument_name: Optional[str] = None
    property_address: Optional[str] = None


class PaginatedObligations(BaseModel):
    items: list[ObligationWithComputed]
    total: int
    page: int
    per_page: int
    pages: int


class ObligationGrouped(BaseModel):
    critical: list[ObligationWithComputed] = []
    high: list[ObligationWithComputed] = []
    normal: list[ObligationWithComputed] = []


class UpdateObligationRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    amount: Optional[Decimal] = None
    next_due: Optional[date] = None
    status: Optional[str] = None
    severity: Optional[str] = None
    alert_days_before: Optional[list[int]] = None


class ObligationCompleteRequest(BaseModel):
    payment_amount: Optional[Decimal] = None
    payment_date: Optional[date] = None
    payment_method: Optional[str] = None


# ---------------------------------------------------------------------------
# Payment — paginated, summary
# ---------------------------------------------------------------------------

class PaginatedPayments(BaseModel):
    items: list[PaymentResponse]
    total: int
    page: int
    per_page: int
    pages: int


class PropertyPaymentSummary(BaseModel):
    property_id: UUID
    address: str
    total: float


class MonthlyHistory(BaseModel):
    month: str
    incoming: float
    outgoing: float
    net: float


class PaymentSummaryResponse(BaseModel):
    total_outgoing_monthly: float
    total_incoming_monthly: float
    net_monthly: float
    outgoing_by_property: list[PropertyPaymentSummary] = []
    incoming_by_property: list[PropertyPaymentSummary] = []
    payment_history: list[MonthlyHistory] = []


# ---------------------------------------------------------------------------
# Dashboard
# ---------------------------------------------------------------------------

class BalloonAlert(BaseModel):
    instrument_name: str
    property_address: str
    balloon_date: Optional[date] = None
    balloon_amount: Optional[Decimal] = None
    days_until: Optional[int] = None


class WrapSpreadItem(BaseModel):
    property_address: str
    monthly_spread: float
    annual_spread: float


class DueOnSaleRiskItem(BaseModel):
    property_address: str
    instrument_name: str
    risk_level: str


class FinancingDashboardResponse(BaseModel):
    total_instruments: int = 0
    total_balance: float = 0
    total_monthly_obligations: float = 0
    total_monthly_income: float = 0
    net_monthly_cash_flow: float = 0
    upcoming_balloons: list[BalloonAlert] = []
    wrap_spreads: list[WrapSpreadItem] = []
    due_on_sale_risks: list[DueOnSaleRiskItem] = []
